### ISSUE TITLE
Add: CIFAR10-DVS dataset

### ISSUE THREAD

**Issue by MrPeterJin** [neuromorphs/tonic#170]
Hi,
Firstly thanks for your work on this package. I wonder that if you have plans to add CIFAR10-DVS dataset into your Tonic.dataset class?

**Comment by biphasic** [comment]
Hey,
thanks for your message. Adding this to the todo list. At the moment I'm a bit tight on time, so a pull request would be greatly appreciated. I have some notes on that [here](https://tonic.readthedocs.io/en/latest/about/contribute.html#how-can-i-add-a-new-dataset)

**Comment by MrPeterJin** [comment]
Thanks for your documentation, will try!

**Comment by MrPeterJin** [comment]
> Hey, thanks for your message. Adding this to the todo list. At the moment I'm a bit tight on time, so a pull request would be greatly appreciated. I have some notes on that [here](https://tonic.readthedocs.io/en/latest/about/contribute.html#how-can-i-add-a-new-dataset)

Hi, I have progressed on CIFAR10-DVS dataset. However, there is one problem I am working on. The data in CIFAR10-DVS is in `aedat 2.0` format, which I am investigating to obtain the events. I see in your package, you have the function `read_aedat4` for `aedat 4.0` data. Therefore I would like to ask if you have any solutions for converting the 2.0 format to 4.0?

**Comment by tobidelbruck** [comment]
I made the tool aedat4to2 (https://github.com/SensorsINI/aedat4to2) but 
this is dual problem, is that correct?

Maybe it is not so hard?

**Comment by MrPeterJin** [comment]
> I made the tool aedat4to2 (https://github.com/SensorsINI/aedat4to2) but this is dual problem, is that correct? Maybe it is not so hard?

Yeah, that's correct. I've seen your repo and currently I am investigating `dv-python` package for a direct reading. Thanks for the comment.

**Comment by MrPeterJin** [comment]
I think I need help with CIFAR10-DVS or neuromorphic data insights. When I check raw results from `NMNIST` datasets, it returns a ndarray like:
 `[(  6, 139,   63, 1) (222, 147,  140, 1) (170,  65,  177, 0), (148, 115,  180, 0) ( 97, 127,  206, 1) (174,  43,  257, 1), (182,  13,  289, 0)...`.

However, my investigations with `dv-python` for CIFAR10-DVS only reads:
 `[(6, 0, 0, -1), (4, 0, 1, -1), (4, 0, 1, -1), (5, 0, 4, -1), (5, 0, 4, -1), (5, 0, 5, -1), (0, 0, 5, -1), (3, 0, 5, -1), (2, 0, 6, 1), (2, 0, 6, -1), (2, 0, 7, -1), (1, 0, 7, -1), (1, 0, 7, -1), (6, 0, 8, -1),...`,
which is way smaller than the `NMNIST` dataset. My implementation at the very first stage is described as follows, and I am sure that I missed something like post-processing or so:

```python
    def __getitem__(self, index):
        """
        Returns:
            a tuple of (events, target) where target is the index of the target class.
        """
        events = read_aedat2(self.data[index]) # this reads the raw data
        target = self.targets[index]

        if self.transform is not None:
            events = self.transform(events)
        if self.target_transform is not None:
            target = self.target_transform(target)
        return events, target
```

In the `read_aedat2` function:

```python

def read_aedat2(in_file):
    from dv import LegacyAedatFile
    """
    Get the aer events from version 2 of .aedat file

    Args:
        in_file: str The name of the .aedat file
    Returns:
        xytp:   numpy structured array of events
    """
    with LegacyAedatFile(in_file) as f:
        xytp = []
        for event in f:
            x, y, t = event.x, event.y, event.timestamp
            if event.polarity == True:
                p = 1
            else:
                p = -1
            xytp.append((x, y, t, p))

    return xytp
```

Again, this development is at the very first stage. Some format alignment (like ndarray) is not implemented yet. Also, I am a newcomer to neuromorphic datasets. So I think maybe someone can give some background knowledge about the CIFAR10-DVS dataset?

**Comment by MrPeterJin** [comment]
> I think I need help with CIFAR10-DVS or neuromorphic data insights. When I check raw results from `NMNIST` datasets, it returns a ndarray like: `[( 6, 139, 63, 1) (222, 147, 140, 1) (170, 65, 177, 0), (148, 115, 180, 0) ( 97, 127, 206, 1) (174, 43, 257, 1), (182, 13, 289, 0)...`.
> 
> However, my investigations with `dv-python` for CIFAR10-DVS only reads: `[(6, 0, 0, -1), (4, 0, 1, -1), (4, 0, 1, -1), (5, 0, 4, -1), (5, 0, 4, -1), (5, 0, 5, -1), (0, 0, 5, -1), (3, 0, 5, -1), (2, 0, 6, 1), (2, 0, 6, -1), (2, 0, 7, -1), (1, 0, 7, -1), (1, 0, 7, -1), (6, 0, 8, -1),...`, which is way smaller than the `NMNIST` dataset. My implementation at the very first stage is described as follows, and I am sure that I missed something like post-processing or so:
> 
> ```python
>     def __getitem__(self, index):
>         """
>         Returns:
>             a tuple of (events, target) where target is the index of the target class.
>         """
>         events = read_aedat2(self.data[index]) # this reads the raw data
>         target = self.targets[index]
> 
>         if self.transform is not None:
>             events = self.transform(events)
>         if self.target_transform is not None:
>             target = self.target_transform(target)
>         return events, target
> ```
> 
> In the `read_aedat2` function:
> 
> ```python
> def read_aedat2(in_file):
>     from dv import LegacyAedatFile
>     """
>     Get the aer events from version 2 of .aedat file
> 
>     Args:
>         in_file: str The name of the .aedat file
>     Returns:
>         xytp:   numpy structured array of events
>     """
>     with LegacyAedatFile(in_file) as f:
>         xytp = []
>         for event in f:
>             x, y, t = event.x, event.y, event.timestamp
>             if event.polarity == True:
>                 p = 1
>             else:
>                 p = -1
>             xytp.append((x, y, t, p))
> 
>     return xytp
> ```
> 
> Again, this development is at the very first stage. Some format alignment (like ndarray) is not implemented yet. Also, I am a newcomer to neuromorphic datasets. So I think maybe someone can give some background knowledge about the CIFAR10-DVS dataset?
> 
> By the way, the link for the original dataset is https://figshare.com/articles/dataset/CIFAR10-DVS_New/4724671/2

I worked out how to solve it. Will PR soon.

**Comment by biphasic** [comment]
Hey,
sorry for the late reply. Many thanks for looking into this. I suggest that you open up a PR so that we can look at this together. I'll be able to comment on stuff and you can still push new commits to that same branch. The dv package is quite a large dependency so I'm not sure I want to include it as a global dependency, but I understand that there seems to be no other easy way to read AEDAT 2 files. @tobidelbruck  if you put your package on PyPI we could potentially rely on yours instead. For the moment let's keep it a local dependency, meaning that if someone wants to use CIFAR10-DVS, then they need dv installed, otherwise they don't. Waiting for your PR (even if not complete yet) so that I can take a look. Thanks!

**Comment by MrPeterJin** [comment]
> open up a PR so that we can look at this together. I'll be able to comment on stuff and you can still push new commits to that same branch. The dv package is quite a large dependency so I'm not sure I want to include it as a global dependency, but I understand that there seems to be no other easy way to read AEDAT 2 files. @tobidelbruck if you put your package on PyPI we could potentially rely on yours instead. For the moment let's keep it a local dependency, meaning that if someone wants to use CIFAR10-DVS, then they need dv installed, otherwise they don't. Waiting for your PR (even if not complete yet) so that I can take a look. Thanks!

Thanks for your comment. One thing I want to clear out is that the user does not require to install `dv` because the original dataset is already be converted. That means it is a converted version of the original dataset, and I have uploaded it on my Google Drive. This is also one of my concerns. Do you have any recommendations for cloud storage that can hold datasets for a reliable long time?

**Comment by biphasic** [comment]
That's nice, thank you for that. Your question is difficult to answer, as free hosters such as Dropbox and Google Drive are reliable in terms of uptime but depending on the user's subscription they might block access after a certain number of downloads per day (which has happened with NMNIST and a free Dropbox plan for example). CIFAR10-DVS being not as popular, we'll probably get away with it. I have done the same thing you did and pre-processed the DVS Gesture dataset and uploaded it on my old lab's server. But for those cases often the bandwidth is limited. There exist alternative methods such as IPFS as suggested in https://github.com/neuromorphs/tonic/issues/116 but I don't know how much of a long-term solution that would be. For the moment GDrive should be fine. Maybe you can also make public your conversion scripts, that way someone else could potentially replicate the same data. 
On another note you might find some useful functions to download from Google Drive [here](https://github.com/pytorch/vision/blob/main/torchvision/datasets/utils.py)

**Comment by MrPeterJin** [comment]
> That's nice, thank you for that. Your question is difficult to answer, as free hosters such as Dropbox and Google Drive are reliable in terms of uptime but depending on the user's subscription they might block access after a certain number of downloads per day (which has happened with NMNIST and a free Dropbox plan for example). CIFAR10-DVS being not as popular, we'll probably get away with it. I have done the same thing you did and pre-processed the DVS Gesture dataset and uploaded it on my old lab's server. But for those cases often the bandwidth is limited. There exist alternative methods such as IPFS as suggested in #116 but I don't know how much of a long-term solution that would be. For the moment GDrive should be fine. Maybe you can also make public your conversion scripts, that way someone else could potentially replicate the same data. On another note you might find some useful functions to download from Google Drive [here](https://github.com/pytorch/vision/blob/main/torchvision/datasets/utils.py)

Thanks for your description. Currently I am testing if the download link can maintain stability (you know Google Drive would pop up virus scan page while a file is too large, so I have to get a direct link). Once this is done, I would PR as soon as possible.

**Comment by MrPeterJin** [comment]
Hi, just let you know that I would PR in 6 hrs. For storage, I have created a Google Drive link using its API. However due to extensive testing it is out of download quota now... Also I have tried MEGA cloud storage but there seemed isn't a way to generate  a direct download link (the converted file is ~10GB). Anyway I would upload it as soon as possible.

**Comment by biphasic** [comment]
merged https://github.com/neuromorphs/tonic/pull/172

**Comment by biphasic** [comment]
Hey @MrPeterJin, I added your dataset to the [documentation](https://tonic.readthedocs.io/en/latest/reference/datasets.html#cifar10-dvs) and also included it in the latest release, 1.0.13! Thanks again :)
